### PR TITLE
Add User-Agent header to Java client

### PR DESCRIPTION
**Pull request by mikaelm12**
Issue:#12295 
It looks like we already add the User-Agent header in the .NET Client and since the header is set by browsers by default we only need to set it in the Java client now.

https://github.com/aspnet/AspNetCore/blob/8c02467b4a218df3b1b0a69bceb50f5b64f482b1/src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs#L554

### CODE REVIEW

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java [hunk 1]
```diff
@@ -327,6 +327,7 @@ public Completable start() {
         handshakeResponseSubject = CompletableSubject.create();
         handshakeReceived = false;
         CompletableSubject tokenCompletable = CompletableSubject.create();
+        localHeaders.put("User-Agent", "Microsoft/SignalR/3.1/OtherUserAgentThings");
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java]: Well *that's* definitely not going to be the user agent ;P.

Comment (anurse) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java]: Perhaps this? Where `version` is a value get from a constant generated by the build or something.

```suggestion
        localHeaders.put("User-Agent", "Microsoft SignalR/" + version + " (Java)");
```

**Review by analogrelay — COMMENTED** [review]
A few things.

If it's for 3.1 and there's time, we'll do it in 3.0 :). This is motivated by Azure SignalR wanting this info, so it seems OK. If we can land this change this week, I don't see a concern with taking this in 3.0. It depends on what we can figure out from the below discussion though :). Let's get it right and then figure out what vehicle it will fit in.

Let's make the user agent values consistent across as many clients as we can. Right now, the .NET client's user agent is `Microsoft.AspNetCore.Http.Connections.Client/[version]`. Yuck. The layering makes sense, but I don't like it. My vote would be to just compromise the layering a bit here and call it "Microsoft SignalR". It's just a string, nothing actually couples the layers by doing this. A huge majority of people using the lower layer will be using "SignalR" as we know it.

My proposed format for the User-Agent string is something like this:

```
Microsoft SignalR/[Version] ([Detailed Version]; [Operating System]; [Runtime]; [Runtime Version])
```

Where:
* `[Version]` is the major/minor of the SignalR client itself.
  * For .NET client, we can just use the major and minor components of the Assembly version.
  * For JS we have a `VERSION` constant which is the full version. We can parse it, or add a new `SHORT_VERSION` constant or something with just the major/minor component.
  * For Java I don't know, we'll have to find out.
* `[Detailed Version]` is the really detailed version (including full build number). In .NET we store this in the `AssemblyInformationalVersionAttribute`. TBD for Java/JS
* `[Operating System]` is an OS type, just based on [`RuntimeInformation.IsOSPlatform`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.isosplatform?view=netframework-4.8) in .NET, `process.platform` in JS.
	* `Windows NT` for Windows
	* `macOS` for macOS
	* `Linux` for Linux
* `[Runtime]` is a simple name defining the kind of runtime the client is on (`.NET`, `Java`, `NodeJS` would be the three options we have right now)
* `[Runtime Version]` is a version of the runtime (.NET, Java, NodeJS, etc.):
	* .NET: [`RuntimeInformation.FrameworkDescription`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.frameworkdescription?view=netframework-4.8) is probably best
	* Java: Probably the [`java.version` system property](https://docs.oracle.com/javase/tutorial/essential/environment/sysprop.html), maybe with `java.vendor` prepended?
	* NodeJS: `process.versions.node`

Examples:

* .NET Core: `Microsoft SignalR/3.0 (3.0.0-preview9.12345.1; Windows NT; .NET; .NET Core 4.6.27317.03)` (the version number FrameworkDescription spits out is a bit odd but we can back-trace it to the "marketing" version)
* Java: `Microsoft SignalR/3.0 (3.0.0-preview9.12345.1; macOS; Java; 1.5.0)`
* NodeJS: `Microsoft SignalR/3.0 (3.0.0-preview9.12345.1; Linux; NodeJS; 11.11.0)`

So there are two work items here I think:

1. Figure out how to properly get the version of the Java client from **within the Java client itself*. We may actually have to just generate a Java class with it as a constant, if there's no built-in way (like our attributes in .NET).

1. Update all the user agents to have this data.

And of course, a user-specified `User-Agent` value should always override.

Thoughts?

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java (hunk 1 above)
Comment (mikaelm12) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java]: > Well that's definitely not going to be the user agent ;P

I remember hearing stories of `ICanHas` prefixes lol. But yeah, I meant to comment to ask for what we wanted here

Comment (anurse) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java]: Oh I ended up with comments. I ended up with more comments that I originally thought (see above) ;P.

**Comment by BrennanConroy** [comment]
We should also take a look at https://github.com/aspnet/AspNetCore/blob/8c02467b4a218df3b1b0a69bceb50f5b64f482b1/src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs#L40 and see if we can fix it now

**Comment by halter73** [comment]
> For JS we have a VERSION constant which is the full version. We can parse it, or add a new SHORT_VERSION constant or something with just the major/minor component.

How are we planning on sending the user-agent with the JS client? I don't think we can reliably modify the User-Agent header. It looks like the fetch spec [technically allows it](https://github.com/web-platform-tests/wpt/issues/2042), but this [still-open 4-year-old chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=571722) to support modifying the User-Agent makes me think we shouldn't rely on this. Should we use some other header to send the custom user agent? Or is the default browser User-Agent good enough?

**Comment by mikaelm12** [comment]
> Or is the default browser User-Agent good enough? 

That would let consumers know that the client is using the browser, but we wouldn't get any version information, right? Is that good enough for our purposes? @Tratcher mentioned using `X-User-Agent` as an alternative I believe.

**Comment by anurse** [comment]
Plan: Retarget this to `master`. Get it merged in. Then open issue to backport to 3.1 when available. That way we get it in the code ASAP and don't have to wait for the branches to be in the perfect state.

**Comment by mikaelm12** [comment]
I'll squash my commits and re target on to master. Just wanted to finish up the changes I was working on today here.

**Comment by mikaelm12** [comment]
This is what the user agent header looks like now
				Product - Version - Detailed Version - OS - Java - Java Vendor - Java Version
`Microsoft SignalR/99.99.99; 99.99.99-dev; Windows NT; Java; Oracle Corporation; 10.0.1`

**Comment by mikaelm12** [comment]
Also it looks like the bug associated with being able to set restricted headers, aka User-Agent, in websocket client are still open so we still need to comment out the code that @BrennanConroy pointed out above..

**Comment by mikaelm12** [comment]
Tracked here now: https://github.com/aspnet/AspNetCore/pull/13168

**Pull request closed without merging**